server/src/main/java/org/opensearch/common/time/JavaDateFormatter.java

```diff
@@ -287,21 +290,33 @@ private TemporalAccessor doParse(String input) {
         if (parsers.size() > 1) {
             Object object = null;
             DateTimeFormatter lastParsedformatter = null;
-            for (DateTimeFormatter formatter : parsers) {
-                ParsePosition pos = new ParsePosition(0);
-                object = formatter.toFormat().parseObject(input, pos);
-                if (parsingSucceeded(object, input, pos)) {
-                    lastParsedformatter = formatter;
-                    break;
+            parserLock.readLock().lock();
```

Comment:
@CaptainDredge That should make it, but here is the deal: we would pay the cost of synchronization on every `doParse` call, would it wipe out the gains you've benchmarked?

In any case, I would like to run one idea over you: what if we do not change parsers list at all, but instead just cache the `lastParsedformatter` in a volatile field? The `doParse` just check if `lastParsedformatter` is not `null` and try to parse with it. We should not care much about synchronizing `lastParsedformatter` since our assumption is that for same index, even if called concurrently, the parser should pick the same formatter.

Reply:
> That should make it, but here is the deal: we would pay the cost of synchronization on every doParse call, would it wipe out the gains you've benchmarked?

Good question, I need to benchmark once again to see the gains now

> In any case, I would like to run one idea over you: what if we do not change parsers list at all, but instead just cache the lastParsedformatter in a volatile field? The doParse just check if lastParsedformatter is not null and try to parse with it. We should not care much about synchronizing lastParsedformatter since our assumption is that for same index, even if called concurrently, the parser should pick the same formatter.

I feel that's a nice middle ground where we can preveny any synchronization cost and still get the benefits of caching. I like it. Let me implement it asap

Reply:
Done!!